### PR TITLE
Complete remastering

### PR DESCRIPTION
**Pull request by agroupp**
- implemented own state management engine independent from any 3rd party;
- removed providers;
- added conception of hooks;
- separated reactive part into standalone functions;
- separated Angular part;
- added conception of extenders;
- added conception of modifiers;
- implemented status state behavior with extension;
- implemented storage functionality with a modifier;
- added conception of `useState` for declarative components;

And many other minor fixes and tweaks.
See examples in the development application — reactive and declarative "books" versions.

### CODE REVIEW

File: packages/ark/testing/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# @e-square/ark/testing
```
Comment (ronnetzer) [inline packages/ark/testing/README.md:1]: what do you think of converting ark-testing to its own publishable library instead of a secondary entry point?

File: packages/ark/testing/src/index.ts [hunk 2]
```diff
@@ -1,7 +1,5 @@
-export * from './state-provider.mock';
-export * from './test-state';
-export * from './test-provider';
-export * from './local-storage-spy';
-export * from './store.mock';
 export * from './ark-select.mock';
 export * from './ark-select-status.mock';
+export * from './local-storage-spy';
```
Comment (ronnetzer) [inline packages/ark/testing/src/index.ts:3]: although its not part of the PR, I wonder why this library has a local storage spy? seems a bit out of scope

File: packages/ark/testing/src/store.mock.ts [hunk 3]
```diff
@@ -1,15 +1,14 @@
-import { Observable, ReplaySubject } from 'rxjs';
+import { ArkObserver, ArkStore } from '@e-square/ark';
 
-import { ArkStore } from '../src';
-
-export type StoreMock<State> = Partial<Record<keyof ArkStore<State>, jest.Mock | State | Observable<State>>>;
+export type StoreMock<State> = Partial<Record<keyof ArkStore<State>, jest.Mock | State | ArkObserver<State>[]>>;
 export function createStoreMock<State>(initialState: State): StoreMock<State> {
```
Comment (ronnetzer) [inline packages/ark/testing/src/store.mock.ts:4]: we should think how we can make the testing lib a bit more robust and realistic, for example, change the jest.fn() methods to a proxy for the real implementation.

File: packages/ark/tsconfig.lib.json [hunk 4]
```diff
@@ -5,7 +5,7 @@
     "declaration": true,
     "declarationMap": true,
     "inlineSources": true,
-    "types": [],
+    "types": ["jest"],
```
Comment (ronnetzer) [inline packages/ark/tsconfig.lib.json:8]: is it a good idea to couple our testing library to jest?

File: dev-apps/dev-app/src/app/auth/auth.guard.ts [hunk 5]
```diff
@@ -25,7 +25,7 @@ export class AuthGuard implements CanActivate, CanActivateChild, CanDeactivate<u
     route: ActivatedRouteSnapshot,
     state: RouterStateSnapshot,
   ): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
-    return this.sessionStore.select('user').pipe(map(user => (user ? true : this.router.createUrlTree(['login']))));
+    return this.sessionStore.user$.pipe(map(user => (user ? true : this.router.createUrlTree(['login']))));
```
Comment (ronnetzer) [inline dev-apps/dev-app/src/app/auth/auth.guard.ts:28]: should be with take(1) no? 🙈

File: dev-apps/dev-app/src/app/books/books-declarative/books-declarative.component.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  private readonly setBooks: (books: Book[]) => void;
+  private readonly setPaginationData: (paginationData: PaginationData) => void;
+
+  readonly searchControl = new FormControl<string | null>(null);
+
+  paginationData: () => PaginationData;
+  books: () => Book[];
+
+  constructor(private readonly booksService: BooksService) {
+    const [books, setBooks] = useArkState<Book[]>([], {
```
Comment (ronnetzer) [inline dev-apps/dev-app/src/app/books/books-declarative/books-declarative.component.ts:33]: I honestly don't know how I feel about this syntax in angular 😕 maybe we should reconsider and have another layer that is more specific to angular: `injectArkState()` that returns something like `EventEmitter` or our own version of it that is based on our observer implementation

to be specific my issue is the needed boilerplate in order to assign these methods. also that they're functions in the first place

File: packages/ark/src/lib/angular/angular-component.destroyer.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { injetcViewRef } from './inject-view-ref';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function angularComponentDestroyer(): (cb: () => any) => void {
```
Comment (ronnetzer) [inline packages/ark/src/lib/angular/angular-component.destroyer.ts:4]: fyi angular is not officially supporting this way so we'll need to keep an eye on this if it changes in the future

File: packages/ark/src/lib/angular/ark-select.directive.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
+/* eslint-disable @angular-eslint/directive-class-suffix */
+import { Directive, inject, Input, OnChanges, OnDestroy, TemplateRef, ViewContainerRef } from '@angular/core';
+
+import { ArkStore, SubscriptionDescriptor, UnknownState } from '../entities';
+import { AsyncContext } from './async-context';
+
+/**
+ * Extended version of `ArkAsync` directive.
```
Comment (ronnetzer) [inline packages/ark/src/lib/angular/ark-select.directive.ts:8]: if its an extension, why its not extending it in any way? 😄 maybe we can use hostDirectives for this

File: packages/ark/src/lib/angular/async-context.ts [hunk 9]
```diff
@@ -5,7 +5,7 @@ export class AsyncContext<T> {
     return this._value;
   }
 
-  update(value: T): void {
+  update(value: T | undefined): void {
```
Comment (ronnetzer) [inline packages/ark/src/lib/angular/async-context.ts:8]: because its a publishable library we need to support strict mode and make sure null & undefined are explicitly set
```suggestion
  update(value?: T | null): void {
```

File: packages/ark/src/lib/angular/inject-view-ref.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { ChangeDetectorRef, inject, ViewRef } from '@angular/core';
+
+const ERROR_MESSAGE = 'To use Angular features the store must be created in an injection context such as a constructor';
```
Comment (ronnetzer) [inline packages/ark/src/lib/angular/inject-view-ref.ts:3]: not exactly accurate as it will also throw if you try to create it from a service

File: packages/ark/src/lib/entities/hooks/ark-hooks.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  beforeDestroy: BeforeDestroyInterceptor<State>[];
+}
+
+export interface HasHooks<State> {
+  beforeInit?: BeforeInitInterceptor<State>[];
+  afterInit?: AfterInitInterceptor<State>[];
+  beforeUpdate?: BeforeUpdateInterceptor<State>[];
+  afterUpdate?: AfterUpdateInterceptor<State>[];
+  beforeDestroy?: BeforeDestroyInterceptor<State>[];
+}
```
Comment (ronnetzer) [inline packages/ark/src/lib/entities/hooks/ark-hooks.ts:27]:
```suggestion
export type HasHooks<State> = Partial<ArkHooks<State>>;
```

File: packages/ark/src/lib/entities/hooks/ark-hooks.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+export type ArkStoreModifier<State = unknown> = () => Partial<HasHooks<State>>;
```
Comment (ronnetzer) [inline packages/ark/src/lib/entities/hooks/ark-hooks.ts:29]: Partial is redundant here no? (even without my previous suggestion)
```suggestion
export type ArkStoreModifier<State = unknown> = () => HasHooks<State>;
```

File: packages/ark/src/lib/mixins/store.mixin.ts [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        const destroyer = options.destroyer();
+        destroyer(() => this.destroy());
+      }
+
+      let state =
+        typeof initialState !== 'object'
+          ? initialState
+          : Array.isArray(initialState)
+          ? (initialState.slice(0) as State)
+          : { ...initialState };
```
Comment (ronnetzer) [inline packages/ark/src/lib/mixins/store.mixin.ts:39]: is shallow clone being done on purpose?
also do we want to implement Object.freeze for development mode?

File: packages/ark/src/lib/mixins/store.mixin.ts [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        state = arg.slice(0) as State;
+      } else if (typeof arg === 'object') {
+        if (this._hooks.beforeUpdate.length) {
+          for (const interceptor of this._hooks.beforeUpdate) {
+            interceptor(this.state, arg, this);
+          }
+        }
+
+        state = { ...this.state, ...arg };
+      } else if (typeof arg === 'boolean' || typeof arg === 'number' || typeof arg === 'string') {
```
Comment (ronnetzer) [inline packages/ark/src/lib/mixins/store.mixin.ts:77]: we need a utils lib and extract all of these checks to it

File: packages/ark/src/lib/status/status.mixin.ts [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    constructor(...args: any[]) {
+      super(...args);
+      this.hooks.addBeforeDestroy(() => this.statusStore.destroy());
+    }
+
+    updateStatus(statusState: Partial<StatusState>): void {
+      if (statusState.status === 'error' && !statusState.error) {
+        throw new Error('Setting status to error requires updating the `error` property as well.');
```
Comment (ronnetzer) [inline packages/ark/src/lib/status/status.mixin.ts:18]: not so sure if I'll make it that strict, what will happen if we allow this? the component will be in error state without any message?

File: packages/ark/src/lib/storage/with-storage.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+  return () => ({
+    beforeInit: [
+      initialState => {
+        const valueInStorage = storageService.readStorage()?.[options.name];
+
+        if (!valueInStorage) {
+          return initialState;
+        }
+
+        return { ...initialState, ...valueInStorage };
```
Comment (ronnetzer) [inline packages/ark/src/lib/storage/with-storage.ts:17]: we should allow the user to specify his own merge function

File: packages/ark/src/lib/store.ts [hunk 17]
```diff
@@ -1,29 +1,47 @@
...
+  return new storeConstructor(initialState, options);
+}
+
+/**
+ * Lets you add a state variable to your component.
+ *
+ * @example
+ * const [books, setBooks] = useArkState<Book[]>([]);
+ */
+export function useArkState<State, S extends ArkStore<State> = ArkStore<State>>(
```
Comment (ronnetzer) [inline packages/ark/src/lib/store.ts:40]: I think this should be implemented per framework

**Review by ronnetzer — COMMENTED** [review]
In general, this looks great. 
I think we achieved here a decoupling level that allows us to take it to the next level and create adapters per framework.
I suggest to separate ark into multiple packages: ark-core, ark-testing, and per framework package (ark-angular, ark-react, etc).
this approach will allow us to create more angular-ish API without coupling the core library to the framework which currently forces the API to be less intuitive to angular (useArkState for example)

File: packages/ark/src/lib/status/status.mixin.ts (hunk 15 above)
Comment (agroupp) [reply packages/ark/src/lib/status/status.mixin.ts:18]: yes and I believe there is no sense in error without the message. You can't even create JS `Error` object without the message.

File: packages/ark/src/lib/store.ts (hunk 17 above)
Comment (agroupp) [reply packages/ark/src/lib/store.ts:40]: It's pretty generic, you can use it even in plain JS

File: dev-apps/dev-app/src/app/books/books-declarative/books-declarative.component.ts (hunk 6 above)
Comment (agroupp) [reply dev-apps/dev-app/src/app/books/books-declarative/books-declarative.component.ts:33]: I hate this syntax in Angular and would never use it. I did this example for @OmerGronich 😊

File: packages/ark/src/lib/angular/angular-component.destroyer.ts (hunk 7 above)
Comment (agroupp) [reply packages/ark/src/lib/angular/angular-component.destroyer.ts:4]: This is the only way to destroy functionally created store inside the component

File: packages/ark/src/lib/angular/ark-select.directive.ts (hunk 8 above)
Comment (agroupp) [reply packages/ark/src/lib/angular/ark-select.directive.ts:8]: It's not an extension of asyncPipe, it's its extended version.

File: packages/ark/src/lib/angular/inject-view-ref.ts (hunk 10 above)
Comment (agroupp) [reply packages/ark/src/lib/angular/inject-view-ref.ts:3]: I copied this message with a small adjustment from the Angular source code.

File: packages/ark/src/lib/mixins/store.mixin.ts (hunk 13 above)
Comment (agroupp) [reply packages/ark/src/lib/mixins/store.mixin.ts:39]: Not sure I get which object to freeze

File: packages/ark/testing/README.md (hunk 1 above)
Comment (agroupp) [reply packages/ark/testing/README.md:1]: Furthermore, I plan to split it into at least 3 libs - store itself, ark-rxjs and ark-angular

File: packages/ark/testing/src/index.ts (hunk 2 above)
Comment (agroupp) [reply packages/ark/testing/src/index.ts:3]: this is needed for storage modifier UTs

File: packages/ark/src/lib/mixins/store.mixin.ts (hunk 13 above)
Comment (ronnetzer) [reply packages/ark/src/lib/mixins/store.mixin.ts:39]: take akita for example: https://github.com/salesforce/akita/blob/master/packages/akita/src/lib/store.ts#L180

their internal setState (which is used in update/set/initial state) performs deepFreeze when running in development mode to enforce immutability

File: dev-apps/dev-app/src/app/books/books-declarative/books-declarative.component.ts (hunk 6 above)
Comment (OmerGronich) [reply dev-apps/dev-app/src/app/books/books-declarative/books-declarative.component.ts:33]: I don't like array destructuring for class components

I do like the syntax @eliraneliassy showed from the GDE summit

```
// read
signal() or signal.get()

// write
signal.set(...)
```

**Review by OmerGronich — COMMENTED** [review]
@agroupp 
The latest update to Ark is truly impressive, I'm still going through the code but wanted to share some first impressions.

I like the new approach to store extension and modification, as well as the addition of hooks to further enhance this. 

One thing I wanted to bring up is that since Angular is introducing a new reactive primitive to the framework, it could be worth reconsidering the state implementation internally within Ark. 

It might be more beneficial to create a wrapper for the Angular reactive primitive, similar to [Pinia.js](https://pinia.vuejs.org/introduction.html) for Vue. This wrapper would not change the syntax for using Angular primitives it will only add support for dev tools, SSR and other plugins like ORM/entities/resource etc...

Possible syntax for this could look like this: 
```
import { signal, computed, effect } from '@angular/core'

const useTodosStore = defineArkStore('todosStore', () => {
   const todos = signal([]);
   const completedTodos = computed(() => todos().filter(todo => todo.completed));
   
   effect(() => {
     localStorage.set(todos())
   })
   
   return {
    todos,
    completedTodos
   }
})

@Component(...)
class TodosComponent {
 todosStore = useTodosStore();
}
```

This would just enhance the primitive, not require the user to learn any new syntax just for our library. 

This syntax could be made more "Angularian" if our `defineStore` function returns a structure similar to all providers like so:

```
import { signal, computed, effect } from '@angular/core'

// provider structure: { provide: TODOS_STORE, useFactory: ... }
const todosStoreProvider = defineArkStore('todosStore', () => {
   const todos = signal([]);
   const completedTodos = computed(() => todos().filter(todo => todo.completed));
   
   effect(() => {
     localStorage.set(todos())
   })
   
   return {
    todos,
    completedTodos
   }
})

const TODOS_STORE = getTokenFromProvider(todosStoreProvider);

@Component(
 ...
 providers: [todosStoreProvider]
...
 )
class TodosComponent {
 todosStore = inject(TODOS_STORE);
}
```
